Complete MAIR_EL1

erikjv opened the pull request.

erikjv committed:
Complete MAIR_EL1

andre-richter commented:
Hi,

thanks for taking on this big chunk. Seeing your change made me reflect about the naming scheme I had started here.

Unless explicitly imported with a `use` statement (which I deem an unusual case), the constants would appear in the source code with their module prepended. But then, there is some redundancy. For example:

```rust
MAIR_EL1::Attr0_HIGH::Device + MAIR_EL1::Attr0_LOW_DEVICE::Device_nGnRE,
```

`Attr0_HIGH::Device` makes sense. `Attr0_LOW_DEVICE::Device_nGnRE` on the other hand, has "device" in there two times. Maybe not needed.

Or `Attr0_LOW_MEMORY::Memory_InnerWriteThrough_Transient`. Two times memory.

So in my opinion, for the `_HIGH` attribute, it makes sense to have the `Memory_` prepended. But we could spare the `Device_` for `AttrX_LOW_DEVICE` and the `Memory_` for `AttrX_LOW_MEMORY`.

What do you think?

alilee commented:
For the sake of brevity/clarity, would you consider:
```rust
MAIR_EL1::Attr0_HIGH::Device + 
MAIR_EL1::Attr0_LOW_DEVICE::nGnRE
```
and
```rust
MAIR_EL1::Attr0_HIGH::Outer_WriteThrough_Transient_RWAlloc +
MAIR_EL1::Attr0_LOW_MEMORY::Inner_WriteThrough_Transient
```

alilee commented:
Also, I accept what you say about the use being an unusual case, but the way I'd want to write this register is:
```rust
use MAIR_EL1::ATTRS::*;
MAIR_EL1.write(
    A0_Device + A0_nGnRE +
    A1_Outer_WriteThrough_Transient_RWAlloc + A1_Inner_WriteThrough_Transient +
    A2_Outer_WriteThrough_Transient_RAlloc  + A2_Inner_WriteThrough_Transient
);
```

alilee commented:
And, if you were up for an alternate top-level, do you like this:
```rust
use MAIR_EL1::*;
MAIR_EL1.write(
    Attr0_Device::nGnRE +
    Attr1_Normal_Outer::WriteThrough_Transient_RWAlloc + Attr1_Normal_Inner::WriteThrough_Transient
);
```

erikjv commented:
> And, if you were up for an alternate top-level, do you like this:

I added a patch to try out how that reads, but only for `Attr0`/`Attr1` for now. Let me know what you think. I'll change the  other `Attr`s when this is the right approach.

I also added a `fn tst()` to see how it looks (well, it's just a copy of Alister's code.

alilee commented on src/regs/mair_el1.rs:
```diff
@@ -322,105 +322,105 @@ register_bitfields! {u64,
-            Device_nGnRE = 0b0100,
-            Device_nGRE = 0b1000,
-            Device_GRE = 0b1100
+        Attr1_Device OFFSET(8) NUMBITS(4) [
```
The Device items have to be 8 bits wide, not 4.

alilee commented on src/regs/mair_el1.rs:
```diff
@@ -322,105 +322,105 @@ register_bitfields! {u64,
-            Memory_OuterWriteThrough_Transient_WriteAlloc = 0b0001,
-            Memory_OuterWriteThrough_Transient_ReadAlloc = 0b0010,
-            Memory_OuterWriteThrough_Transient_ReadAlloc_WriteAlloc = 0b0011,
+            WriteThrough_Transient_WAlloc = 0b0001,
```
Aarch64 ignores transient on Outer so i didn't code these. Not sure if that's the right answer - they are defined.

alilee commented on src/regs/mair_el1.rs:
```diff
@@ -322,105 +322,105 @@ register_bitfields! {u64,
+            WriteBack_NonTransient = 0b1100,
+            WriteBack_NonTransient_WAlloc = 0b1101,
+            WriteBack_NonTransient_RAlloc = 0b1110,
+            WriteBack_NonTransient_RWAlloc = 0b1111
```
When I wrote mine (which I've pushed to my repo, btw), I liked:
`WriteBack_NonTransient_AllocRW`
rather than:
`WriteBack_NonTransient_RWAlloc`

andre-richter replied:
In the AArch64 ref manual they write it as `Write-Allocate policy` and `Read-Allocate policy`.
Therefore, I'd prefer to just stick to that ordering of words.

Also, I'd like to stick to the policy of this crate to be as elaborate as possible in the naming of constants (I didn't even follow through myself with the nGnRnE, for example). They are not programmed that often, so if the programming spans a few lines of source code, that's okay as long as it removes any last doubt of what's happening.

I like your ideas of having `Attr0_Device` and `Attr0_Normal_xx`, that makes it more concise!

Here's my proposal:

```rust
        Attr0_Device OFFSET(0) NUMBITS(8) [
            nonGathering_nonReordering_noEarlyWriteAck = 0b00000000,
            nonGathering_nonReordering_EarlyWriteAck = 0b00000100,
            nonGathering_Reordering_EarlyWriteAck = 0b00001000,
            Gathering_Reordering_EarlyWriteAck = 0b00001100

...

        Attr0_Normal_Inner OFFSET(0) NUMBITS(4) [
            WriteThrough_Transient = 0x0000,
            WriteThrough_Transient_WriteAlloc = 0x0001,
            WriteThrough_Transient_ReadAlloc = 0x0010,
            WriteThrough_Transient_ReadWriteAlloc = 0x0011,
```

erikjv replied:
About Device: I prefer the shorter notation (`nGnRnE` & friends), as it's used in other code-bases, and in the documentation (e.g. any use of `DeviceType` (J1-5398)). Then again, the allocation `MemHint`s are not really readable (J1-5397, `Allocation`). So I'm not really sure.

andre-richter replied:
Theoretically, we could have both.
Not sure if we should go down this path, though.

andre-richter commented:
Thanks for your work on this so far! I left a comment inline.

erikjv commented:
Finally found some time to clean-up this patch.

andre-richter commented:
Thanks.

Will hopefully have time to skim through it and release it over the weekend.

alilee commented:
Beautiful - thanks erik.

andre-richter commented:
Thanks for the patience with me.

andre-richter commented:
This is released in 3.0.0 now.
